# Detect liver disease patient with decision tree classifier in python
08-detect-liver-disease-patient-w-dt-in-python

In [1]:
# Setup libraries
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

In [2]:
# Read data into dataframe
df = pd.read_csv('../00-Datasets/liver.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 583 entries, 0 to 582
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         583 non-null    int64  
 1   Gender                      583 non-null    object 
 2   Total_Bilirubin             583 non-null    float64
 3   Direct_Bilirubin            583 non-null    float64
 4   Alkaline_Phosphotase        583 non-null    int64  
 5   Alamine_Aminotransferase    583 non-null    int64  
 6   Aspartate_Aminotransferase  583 non-null    int64  
 7   Total_Protiens              583 non-null    float64
 8   Albumin                     583 non-null    float64
 9   Albumin_and_Globulin_Ratio  579 non-null    float64
 10  Dataset                     583 non-null    int64  
dtypes: float64(5), int64(5), object(1)
memory usage: 50.2+ KB


In [4]:
df.head()

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
0   65  Female              0.7               0.1                   187   
1   62    Male             10.9               5.5                   699   
2   62    Male              7.3               4.1                   490   
3   58    Male              1.0               0.4                   182   
4   72    Male              3.9               2.0                   195   

   Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
0                        16                          18             6.8   
1                        64                         100             7.5   
2                        60                          68             7.0   
3                        14                          20             6.8   
4                        27                          59             7.3   

   Albumin  Albumin_and_Globulin_Ratio  Dataset  
0      3.3                        0.90        1  
1      3.2                        0.74        1  
2      3.3                        0.89        1  
3      3.4                        1.00        1  
4      2.4                        0.40        1

In [5]:
df.tail()

Age Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
578   60   Male              0.5               0.1                   500   
579   40   Male              0.6               0.1                    98   
580   52   Male              0.8               0.2                   245   
581   31   Male              1.3               0.5                   184   
582   38   Male              1.0               0.3                   216   

     Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
578                        20                          34             5.9   
579                        35                          31             6.0   
580                        48                          49             6.4   
581                        29                          32             6.8   
582                        21                          24             7.3   

     Albumin  Albumin_and_Globulin_Ratio  Dataset  
578      1.6                        0.37        2  
579      3.2                        1.10        1  
580      3.2                        1.00        1  
581      3.4                        1.00        1  
582      4.4                        1.50        2

In [6]:
df.sample(6)

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
134   18    Male              1.8               0.7                   178   
576   32    Male             15.0               8.2                   289   
552   45    Male              0.6               0.1                   196   
208   70  Female              0.9               0.3                   220   
171   44  Female              1.9               0.6                   298   
261   33    Male              1.5               7.0                   505   

     Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
134                        35                          36             6.8   
576                        58                          80             5.3   
552                        29                          30             5.8   
208                        53                          95             6.1   
171                       378                         602             6.6   
261                       205                         140             7.5   

     Albumin  Albumin_and_Globulin_Ratio  Dataset  
134      3.6                        1.10        1  
576      2.2                        0.70        1  
552      2.9                        1.00        1  
208      2.8                        0.68        1  
171      3.3                        1.00        1  
261      3.9                        1.00        1

In [7]:
# Check for missing values
df.isna().sum()

Age                           0
Gender                        0
Total_Bilirubin               0
Direct_Bilirubin              0
Alkaline_Phosphotase          0
Alamine_Aminotransferase      0
Aspartate_Aminotransferase    0
Total_Protiens                0
Albumin                       0
Albumin_and_Globulin_Ratio    4
Dataset                       0
dtype: int64

In [8]:
# Check for duplicate values
df.duplicated().sum()

13

In [13]:
# Explore some features
df['Dataset'].value_counts()

1    416
2    167
Name: Dataset, dtype: int64

In [15]:
df['Gender'].value_counts()

Male      441
Female    142
Name: Gender, dtype: int64

In [16]:
# Explore tabular summary
df.describe()

Age  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
count  583.000000       583.000000        583.000000            583.000000   
mean    44.746141         3.298799          1.486106            290.576329   
std     16.189833         6.209522          2.808498            242.937989   
min      4.000000         0.400000          0.100000             63.000000   
25%     33.000000         0.800000          0.200000            175.500000   
50%     45.000000         1.000000          0.300000            208.000000   
75%     58.000000         2.600000          1.300000            298.000000   
max     90.000000        75.000000         19.700000           2110.000000   

       Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
count                583.000000                  583.000000      583.000000   
mean                  80.713551                  109.910806        6.483190   
std                  182.620356                  288.918529        1.085451   
min                   10.000000                   10.000000        2.700000   
25%                   23.000000                   25.000000        5.800000   
50%                   35.000000                   42.000000        6.600000   
75%                   60.500000                   87.000000        7.200000   
max                 2000.000000                 4929.000000        9.600000   

          Albumin  Albumin_and_Globulin_Ratio     Dataset  
count  583.000000                  579.000000  583.000000  
mean     3.141852                    0.947064    1.286449  
std      0.795519                    0.319592    0.452490  
min      0.900000                    0.300000    1.000000  
25%      2.600000                    0.700000    1.000000  
50%      3.100000                    0.930000    1.000000  
75%      3.800000                    1.100000    2.000000  
max      5.500000                    2.800000    2.000000

In [17]:
# Prepare dataset
df.dropna(inplace=True)

le = LabelEncoder()
df['Gender'] = le.fit_transform(df['Gender'])

In [20]:
# Split dataset into 80% train and 20% test
X = df.drop('Dataset', axis=1)
y = df['Dataset']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [22]:
# Build model
classifier = DecisionTreeClassifier()
classifier.fit(X_train, y_train)

DecisionTreeClassifier()

In [24]:
# Feed test data into model
y_predict = classifier.predict(X_test)

actual_vs_predict = pd.DataFrame({'Actual': y_test,
                              'Predict': y_predict})
actual_vs_predict.sample(12)

Actual  Predict
165       1        1
17        2        1
348       1        1
14        1        1
319       2        1
561       1        1
316       2        2
324       1        1
501       1        1
46        1        2
31        1        1
357       1        1

In [25]:
# Evaluate the model
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

           1       0.75      0.80      0.78        82
           2       0.43      0.35      0.39        34

    accuracy                           0.67       116
   macro avg       0.59      0.58      0.58       116
weighted avg       0.66      0.67      0.66       116



## Notes
* We wanted to detect patients with liver disease.
* Our dataset had 583 records with 11 features. 5 of which are of data type float, 5 of type integer, and 1 of type object.
* A quick look showed that there were 4 missing values. We dropped the records with missing values. 
* The dataset was split into 80% train and 20% test sets. No validation set was created or used. No hyperparameter tuning occurred.
* We built, trained, and tested a decision tree classifier.
* We evaluated our model using ground truth and several metrics including precision, recall, and f1-score.